File: src/motifhorde/external.py
"""Small helpers for resolving and running external commands."""

from __future__ import annotations

import os
import shutil
import subprocess

DEFAULT_MEME_COMMAND = "meme"
DEFAULT_STREME_COMMAND = "streme"
PACKAGE_BIN_DIR = os.path.join(os.path.dirname(__file__), "bin")
DEFAULT_DIMONT_JAR = os.path.join(PACKAGE_BIN_DIR, "Dimont.jar")
DEFAULT_SLIM_JAR = os.path.join(PACKAGE_BIN_DIR, "SlimDimont.jar")
DEFAULT_BAMM_COMMAND = "BaMMmotif"


def resolve_command(
    command_name: str,
    fallback_path: str | None = None,
    env_var: str | None = None,
) -> str:
    """Resolve a command from environment, PATH, then fallback."""
    if env_var:
        env_value = os.environ.get(env_var)
        if env_value:
            return env_value

    path_value = shutil.which(command_name)
    if path_value:
        return path_value

    return fallback_path or command_name


def resolve_existing_path(
    cli_value: str | None,
    env_var: str,
    default_path: str,
    label: str,
) -> str:
    """Resolve a file path and require that it exists."""
    path = cli_value or os.environ.get(env_var) or default_path
    if not os.path.exists(path):
        raise FileNotFoundError(f"{label} not found: {path}")
    return path


def run_checked(args: list[str], cwd: str | None = None) -> subprocess.CompletedProcess:
    """Run a subprocess and raise a detailed error on failure."""
    result = subprocess.run(args, cwd=cwd, shell=False, capture_output=True, text=True)
    if result.returncode != 0:
        command = " ".join(args)
        raise RuntimeError(
            f"Command failed: {command}\n"
            f"Return code: {result.returncode}\n"
            f"stdout:\n{result.stdout}\n"
            f"stderr:\n{result.stderr}"
        )
    return result

File: src/external.jl
"""Safe resolution and execution of external programs."""

const DEFAULT_MEME_COMMAND = "meme"
const DEFAULT_STREME_COMMAND = "streme"
const DEFAULT_BAMM_COMMAND = "BaMMmotif"

function resolve_command(
    cli_value::Union{Nothing,AbstractString},
    command_name::AbstractString;
    env_var::Union{Nothing,AbstractString}=nothing,
    fallback::Union{Nothing,AbstractString}=nothing,
)
    cli_value !== nothing && !isempty(cli_value) && return String(cli_value)
    if env_var !== nothing
        value = get(ENV, env_var, "")
        !isempty(value) && return value
    end
    path_value = Sys.which(command_name)
    path_value !== nothing && return path_value
    return fallback === nothing ? String(command_name) : String(fallback)
end

function resolve_existing_path(
    cli_value::Union{Nothing,AbstractString},
    env_var::AbstractString,
    default_path::AbstractString,
    label::AbstractString,
)
    path =
        cli_value === nothing ? get(ENV, env_var, String(default_path)) : String(cli_value)
    isfile(path) || throw(ArgumentError("$label not found: $path"))
    return path
end

struct ProcessResult
    args::Vector{String}
    code::Int
    stdout::String
    stderr::String
end

function run_checked(
    args::AbstractVector{<:AbstractString}; cwd::Union{Nothing,AbstractString}=nothing
)
    argv = String.(args)
    isempty(argv) && throw(ArgumentError("external command cannot be empty"))
    command = Cmd(argv)
    cwd === nothing || (command = Cmd(command; dir=String(cwd)))
    stdout_buffer = IOBuffer()
    stderr_buffer = IOBuffer()
    process = run(pipeline(command; stdout=stdout_buffer, stderr=stderr_buffer); wait=false)
    wait(process)
    result = ProcessResult(
        argv, process.exitcode, String(take!(stdout_buffer)), String(take!(stderr_buffer))
    )
    result.code == 0 || throw(
        ErrorException(
            "Command failed (exit code $(result.code)): $(join(result.args, " "))\n" *
            "stdout:\n$(result.stdout)\nstderr:\n$(result.stderr)",
        ),
    )
    return result
end
